DNS error with certain DNS servers that don't handle EDNS

Only on Windows, I often am unable to resolve DNS with Netty

```
Caused by: java.net.UnknownHostException: failed to resolve 'www.googleapis.com' after 3 queries
        at io.netty.resolver.dns.DnsResolveContext.finishResolve(DnsResolveContext.java:845)
        ... 32 more
```

Despite `nslookup www.googleapis.com` for example working fine.

This happens both when invoking commands from the command line in an `msys2` terminal as well as in IntelliJ.

Don't have much else to go on, any recommendations on how I can get some more logs or something to help debug?

--- a/core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java
@@ -16,7 +16,9 @@
 
 package com.linecorp.armeria.client;
 
+import static com.google.common.base.MoreObjects.firstNonNull;
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
 import static io.netty.handler.codec.http2.Http2CodecUtil.DEFAULT_WINDOW_SIZE;
 import static io.netty.handler.codec.http2.Http2CodecUtil.MAX_FRAME_SIZE_LOWER_BOUND;
 import static io.netty.handler.codec.http2.Http2CodecUtil.MAX_FRAME_SIZE_UPPER_BOUND;
@@ -25,19 +27,23 @@ import static java.util.Objects.requireNonNull;
 
 import java.net.InetSocketAddress;
 import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
+import javax.annotation.Nullable;
+
 import com.google.common.base.MoreObjects;
 import com.google.common.base.MoreObjects.ToStringHelper;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 
 import com.linecorp.armeria.common.CommonPools;
 import com.linecorp.armeria.common.Flags;
 import com.linecorp.armeria.common.Request;
-import com.linecorp.armeria.internal.TransportType;
 
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Metrics;
@@ -47,8 +53,7 @@ import io.netty.channel.epoll.EpollChannelOption;
 import io.netty.handler.codec.http2.Http2CodecUtil;
 import io.netty.handler.ssl.SslContextBuilder;
 import io.netty.resolver.AddressResolverGroup;
-import io.netty.resolver.dns.DnsAddressResolverGroup;
-import io.netty.resolver.dns.DnsServerAddressStreamProviders;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
 import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
 
 /**
@@ -73,11 +78,6 @@ public final class ClientFactoryBuilder {
 
     private static final Consumer<SslContextBuilder> DEFAULT_SSL_CONTEXT_CUSTOMIZER = b -> { /* no-op */ };
 
-    private static final Function<EventLoopGroup, DnsAddressResolverGroup>
-            DEFAULT_ADDRESS_RESOLVER_GROUP_FACTORY = eventLoopGroup -> new DnsAddressResolverGroup(
-            TransportType.datagramChannelType(eventLoopGroup),
-            DnsServerAddressStreamProviders.platformDefault());
-
     // Do not accept 1) the options that may break Armeria and 2) the deprecated options.
     @SuppressWarnings("deprecation")
     private static final Set<ChannelOption<?>> PROHIBITED_SOCKET_OPTIONS = ImmutableSet.of(
@@ -91,9 +91,11 @@ public final class ClientFactoryBuilder {
     private boolean shutdownWorkerGroupOnClose;
     private final Map<ChannelOption<?>, Object> channelOptions = new Object2ObjectArrayMap<>();
     private Consumer<? super SslContextBuilder> sslContextCustomizer = DEFAULT_SSL_CONTEXT_CUSTOMIZER;
+    @Nullable
     private Function<? super EventLoopGroup,
-            ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory =
-            DEFAULT_ADDRESS_RESOLVER_GROUP_FACTORY;
+            ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory;
+    @Nullable
+    private List<Consumer<? super DnsNameResolverBuilder>> domainNameResolverCustomizers;
     private int http2InitialConnectionWindowSize = Flags.defaultHttp2InitialConnectionWindowSize();
     private int http2InitialStreamWindowSize = Flags.defaultHttp2InitialStreamWindowSize();
     private int http2MaxFrameSize = Flags.defaultHttp2MaxFrameSize();
@@ -187,12 +189,35 @@ public final class ClientFactoryBuilder {
     /**
      * Sets the factory that creates a {@link AddressResolverGroup} which resolves remote addresses into
      * {@link InetSocketAddress}es.
+     *
+     * @throws IllegalStateException if {@link #domainNameResolverCustomizer(Consumer)} was called already.
      */
     public ClientFactoryBuilder addressResolverGroupFactory(
             Function<? super EventLoopGroup,
                      ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory) {
-        this.addressResolverGroupFactory = requireNonNull(addressResolverGroupFactory,
-                                                          "addressResolverGroupFactory");
+        requireNonNull(addressResolverGroupFactory, "addressResolverGroupFactory");
+        checkState(domainNameResolverCustomizers == null,
+                   "addressResolverGroupFactory() and domainNameResolverCustomizer() are mutually exclusive.");
+        this.addressResolverGroupFactory = addressResolverGroupFactory;
+        return this;
+    }
+
+    /**
+     * Adds the specified {@link Consumer} which customizes the given {@link DnsNameResolverBuilder}.
+     * This method is useful when you want to change the behavior of the default domain name resolver, such as
+     * changing the DNS server list.
+     *
+     * @throws IllegalStateException if {@link #addressResolverGroupFactory(Function)} was called already.
+     */
+    public ClientFactoryBuilder domainNameResolverCustomizer(
+            Consumer<? super DnsNameResolverBuilder> domainNameResolverCustomizer) {
+        requireNonNull(domainNameResolverCustomizer, "domainNameResolverCustomizer");
+        checkState(addressResolverGroupFactory == null,
+                   "addressResolverGroupFactory() and domainNameResolverCustomizer() are mutually exclusive.");
+        if (domainNameResolverCustomizers == null) {
+            domainNameResolverCustomizers = new ArrayList<>();
+        }
+        domainNameResolverCustomizers.add(domainNameResolverCustomizer);
         return this;
     }
 
@@ -348,6 +373,15 @@ public final class ClientFactoryBuilder {
      * Returns a newly-created {@link ClientFactory} based on the properties of this builder.
      */
     public ClientFactory build() {
+        final Function<? super EventLoopGroup,
+                       ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory;
+        if (this.addressResolverGroupFactory != null) {
+            addressResolverGroupFactory = this.addressResolverGroupFactory;
+        } else {
+            addressResolverGroupFactory = new DefaultAddressResolverGroupFactory(
+                    firstNonNull(domainNameResolverCustomizers, ImmutableList.of()));
+        }
+
         return new DefaultClientFactory(new HttpClientFactory(
                 workerGroup, shutdownWorkerGroupOnClose, channelOptions, sslContextCustomizer,
                 addressResolverGroupFactory, http2InitialConnectionWindowSize, http2InitialStreamWindowSize,
@@ -370,6 +404,7 @@ public final class ClientFactoryBuilder {
             EventLoopGroup workerGroup, boolean shutdownWorkerGroupOnClose,
             Map<ChannelOption<?>, Object> socketOptions,
             Consumer<? super SslContextBuilder> sslContextCustomizer,
+            @Nullable
             Function<? super EventLoopGroup,
                      ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory,
             int http2InitialConnectionWindowSize, int http2InitialStreamWindowSize, int http2MaxFrameSize,
@@ -400,7 +435,7 @@ public final class ClientFactoryBuilder {
             helper.add("sslContextCustomizer", sslContextCustomizer);
         }
 
-        if (addressResolverGroupFactory != DEFAULT_ADDRESS_RESOLVER_GROUP_FACTORY) {
+        if (!(addressResolverGroupFactory instanceof DefaultAddressResolverGroupFactory)) {
             helper.add("addressResolverGroupFactory", addressResolverGroupFactory);
         }
 

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultAddressResolverGroupFactory.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultAddressResolverGroupFactory.java
@@ -1,0 +1,54 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import com.linecorp.armeria.internal.TransportType;
+
+import io.netty.channel.EventLoopGroup;
+import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.dns.DnsAddressResolverGroup;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
+import io.netty.resolver.dns.DnsServerAddressStreamProviders;
+
+/**
+ * The default {@link AddressResolverGroup} factory implementation, which enables asynchronous
+ * DNS resolution and tracing by default.
+ */
+final class DefaultAddressResolverGroupFactory
+        implements Function<EventLoopGroup, AddressResolverGroup<InetSocketAddress>> {
+
+    private final Iterable<Consumer<? super DnsNameResolverBuilder>> customizers;
+
+    DefaultAddressResolverGroupFactory(Iterable<Consumer<? super DnsNameResolverBuilder>> customizers) {
+        this.customizers = requireNonNull(customizers, "customizers");
+    }
+
+    @Override
+    public AddressResolverGroup<InetSocketAddress> apply(EventLoopGroup eventLoopGroup) {
+        final DnsNameResolverBuilder nameResolverBuilder = new DnsNameResolverBuilder();
+        nameResolverBuilder.nameServerProvider(DnsServerAddressStreamProviders.platformDefault());
+        nameResolverBuilder.traceEnabled(true);
+        customizers.forEach(customizer -> customizer.accept(nameResolverBuilder));
+        nameResolverBuilder.channelType(TransportType.datagramChannelType(eventLoopGroup));
+        return new DnsAddressResolverGroup(nameResolverBuilder);
+    }
+}

--- a/core/src/test/java/com/linecorp/armeria/client/ClientFactoryBuilderTest.java
+++ b/core/src/test/java/com/linecorp/armeria/client/ClientFactoryBuilderTest.java
@@ -1,0 +1,38 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.Test;
+
+public class ClientFactoryBuilderTest {
+
+    @Test
+    public void addressResolverGroupFactoryAndDomainNameResolverCustomizerAreMutuallyExclusive() {
+        final ClientFactoryBuilder builder1 = new ClientFactoryBuilder();
+        builder1.addressResolverGroupFactory(eventLoopGroup -> null);
+        assertThatThrownBy(() -> builder1.domainNameResolverCustomizer(b -> {}))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessageContaining("mutually exclusive");
+
+        final ClientFactoryBuilder builder2 = new ClientFactoryBuilder();
+        builder2.domainNameResolverCustomizer(b -> {});
+        assertThatThrownBy(() -> builder2.addressResolverGroupFactory(eventLoopGroup -> null))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessageContaining("mutually exclusive");
+    }
+}

--- a/testing-internal/src/main/resources/logback-test.xml
+++ b/testing-internal/src/main/resources/logback-test.xml
@@ -32,6 +32,7 @@
   <logger name="com.linecorp.armeria.logging.traffic.server" level="OFF" />
   <logger name="com.linecorp.armeria.logging.traffic.client" level="OFF" />
   <logger name="com.linecorp.armeria.internal.Http2GoAwayHandler" level="INFO" />
+  <logger name="io.netty.resolver.dns.TraceDnsQueryLifeCycleObserverFactory" level="DEBUG" />
   <logger name="armeria" level="DEBUG" />
   <logger name="loggerTest" level="ALL" additivity="false">
     <appender-ref ref="NOP" />

